### PR TITLE
storage.py: add more logging and an error condition for s3

### LINKED ISSUE #1882: Improve storage-initializer's messages/logs
/kind bug

**Describe the solution you'd like**
Hi everybody! I spent quite some time investigating a configuration issue for the storage-initializer, namely a `STORAGE_URI` env value set like:

```
s3://wmf-ml-models/articlequality/enwiki/20211022183902/model.bin
```
instead of
```
s3://wmf-ml-models/articlequality/enwiki/20211022183902/
```

It was clearly a misconfiguration issue but the log messages from the container made it more confusing:

```
[I 211025 16:39:52 storage-initializer-entrypoint:13] Initializing, args: src_uri [s3://wmf-ml-models/articlequality/enwiki/20211022183902/model.bin] dest_path[ [/mnt/models]
[I 211025 16:39:52 storage:52] Copying contents of s3://wmf-ml-models/articlequality/enwiki/20211022183902/model.bin to local
[I 211025 16:39:52 credentials:1102] Found credentials in environment variables.
[I 211025 16:39:52 storage:85] Successfully copied s3://wmf-ml-models/articlequality/enwiki/20211022183902/model.bin to /mnt/models
```
The above seems to suggest that `model.bin` has been copied successfully, meanwhile what happens is that nothing is copied due to the wrong `STORAGE_URI` (that is an object path, rather than something than contains objects).

Same thing if I put a `STORAGE_URI` that points to no objects or to a non-existent path on s3. I believe that the code in https://github.com/kserve/kserve/blob/master/python/kserve/kserve/storage.py#L114 should check if any object is contained in the URI provided, and report an error if nothing is returned (ideally skipping the "Successfully copied ..."). In this way the pod creation would fail at the storage-initializer bootstrap, and not later on (in my case, the kserve-container failed, rightfully, due to the absence of the model.bin file that it was looking for).

### PR DESCRIPTION
**Pull request by elukey**
The storage initializer for s3 ends up into the following
two use cases:
- If no object is present in the bucket, it doesn't raise
  any error, and later on the kserve-container fails since
  no model binary is present.
- If a full object path is specified (like /a/b/../model.bin)
  then the s3 downloader will create the same path under
  /mnt/models, and if the kserve-container is not smart enough
  it will fail to find the model.bin.
The current storage-initializer logging doesn't give any hint
or errors about the above use case, ending up in difficult-to-debug
scenarios (see the related issue for more info).

Fixes: #1882

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by elukey** [comment]
@yuzisun Hi! Do you think that the PR is eligible for a test?

File: python/kserve/kserve/storage.py [hunk 1]
```diff
@@ -134,6 +135,11 @@ def _download_s3(uri, temp_dir: str):
             if mimetype in ["application/x-tar", "application/zip"]:
                 Storage._unpack_archive_file(target, mimetype, temp_dir)
 
+        # Raise an error if nothing is retrieved from the bucket
+        # provided in input.
+        if count == 0:
+            raise RuntimeError('No object found in %s' % bucket_path)
```
Comment (yuzisun) [inline python/kserve/kserve/storage.py]: Thanks for fixing this! Can you help make the error message consistent with other storage type?

**Comment by yuzisun** [comment]
> * If a full object path is specified (like /a/b/../model.bin)
>   then the s3 downloader will create the same path under
>   /mnt/models, and if the kserve-container is not smart enough
>   it will fail to find the model.bin.

This seems wrong we should fix that, the [code snippet](https://github.com/kserve/kserve/blob/master/python/kserve/kserve/storage.py#L118) here is problematic.

File: python/kserve/kserve/storage.py (hunk 1 above)
Comment (elukey) [reply python/kserve/kserve/storage.py]: Done! Let me know if it is ok now, I tried also to simply the common error msg :)

**Comment by elukey** [comment]
> > * If a full object path is specified (like /a/b/../model.bin)
> >   then the s3 downloader will create the same path under
> >   /mnt/models, and if the kserve-container is not smart enough
> >   it will fail to find the model.bin.
> 
> This seems wrong we should fix that, the [code snippet](https://github.com/kserve/kserve/blob/master/python/kserve/kserve/storage.py#L118) here is problematic.

I thought the same too, and I have updated the patch with a simple fix. I didn't do it before since the logic seems to be repeated elsewhere, for example download_gcs seems to do something similar afaics. At this point we probably need to decide what is the intended semantic for the downloader, and apply it to all occurrences of it.

**Comment by yuzisun** [comment]
@elukey I think the unit test needs a fix as well.

**Comment by elukey** [comment]
> @elukey I think the unit test needs a fix as well.

Definitely, I'll try to fix it. It feels though that the semantic that we are trying to change was intended, afaics the last big change was done in https://github.com/kserve/kserve/pull/1472/files. I am ok to modify the semantic of course, but I am wondering if there are people already relying on it (that we may break with this change).

**Comment by elukey** [comment]
@yuzisun Hi! I fixed the unit test, and also added more comments related to what I think the code does. Let me know!

File: python/kserve/kserve/storage.py [hunk 2]
```diff
@@ -117,16 +117,32 @@ def _download_s3(uri, temp_dir: str):
...
+            # download all the objects listed above, re-creating their subpaths
+            # under the temp_dir.
+            # If 'uri' is set to "s3://test-bucket/a", then the downloader will
+            # add to temp_dir: b/c/model.bin and model.bin.
+            # If 'uri' is set to "s3://test-bucket/a/b/c/model.bin", then
+            # the downloader will add to temp dir: model.bin
+            # (without any subpaths).
             target_key = (
-                obj.key
+                obj.key.rsplit("/", 1)[1]
```
Comment (yuzisun) [inline python/kserve/kserve/storage.py]: would this be an issue when `storageUri` is `s3://test-bucket/model.bin` and `obj.key` is `model.bin`?

Comment (elukey) [reply python/kserve/kserve/storage.py]: @yuzisun good point! I replaced the [1] with [-1], it should work fine even when the rsplit list is one element:
```
>>> "model.bin".rsplit("/",1)[-1]
'model.bin'
>>> "/a/b/c/model.bin".rsplit("/",1)[-1]
'model.bin'
```

Added also a specific unit test for it, lemme know :)

**Comment by yuzisun** [comment]
Thanks @elukey !! This is great and fixed the long standing bug!
 /lgtm
 /approve

**Pull request merged**